jscomp/config_compiler.ml

```diff
@@ -0,0 +1,127 @@
+  function(obj, key) { delete obj[key] }
+|}]
+
+(* need check which variables exist when we update compiler *)
+let map = [%obj {
+  _LIBDIR = "standard_library_default";
+  _BYTERUN = "standard_runtime";
+  _CCOMPTYPE = "ccomp_type";
+  _BYTECC = "bytecomp_c_compiler";
+  _BYTECCLIBS = "bytecomp_c_libraries";
+  _NATIVECC = "native_c_compiler";
+  _NATIVECCLIBS = "native_c_libraries";
+  _PACKLD = "native_pack_linker";
+  _RANLIBCMD = "ranlib";
+  _ARCMD = "ar";
+  _CC_PROFILE = "cc_profile";
+
+  _MKDLL = "mkdll"; (* undefined *)
+  _MKEXE = "mkexe"; (* undefined *)
+  _MKMAINDLL = "mkmaindll"; (* undefined TODO= upstream to print it too *)
+
+  _ARCH = "architecture";
+  _MODEL = "model";
+  _SYSTEM = "system";
+  _ASM = "asm";
+  _ASM_CFI_SUPPORTED = "asm_cfi_supported"; (* boolean *)
+  _WITH_FRAME_POINTERS = "with_frame_pointers"; (* boolean *)
+  _EXT_OBJ = "ext_obj";
+  _EXT_ASM = "ext_asm";
+  _EXT_LIB = "ext_lib";
+  _EXT_DLL = "ext_lib";
+  _HOST = "host";
+  _TARGET = "target";
+  _SYSTHREAD_SUPPORT = "systhread_supported"; (* boolean *)
+}]
+
+let patch_config jscomp_dir config_map is_windows =
+  let whole_compiler_config = Path.join([| jscomp_dir; "bin"; "config_whole_compiler.mlp" |]) in
+  let whole_compiler_config_output = Path.join([| jscomp_dir; "bin"; "config_whole_compiler.ml" |]) in
+  let content = Fs.readFileSync whole_compiler_config `utf8 in
+  let replace_values whole match_ =
+    match (match_, is_windows) with
+      | ("LIBDIR", true) ->
+        "Filename.concat (Filename.concat (Filename.concat (Filename.dirname Sys.executable_name) \"..\") \"lib\") \"ocaml\""
+      | ("LIBDIR", false) ->
+        let origin_path = Path.join [|jscomp_dir; ".."; "lib"; "ocaml"|] in
+        Js.Json.stringify (Js.Json.string origin_path)
+      | _ ->
+        let map_val = Js.Dict.unsafeGet (Js_cast.dictOfObj map) match_ in
+        Js.Dict.unsafeGet config_map map_val
+  in
+  let generated = Js.String.replaceByFun1 (Js.Re.fromStringWithFlags "%%(\\w+)%%" ~flags:"g") replace_values content in
+  Fs.writeFileSync whole_compiler_config_output generated `utf8
+
+let get_config_output is_windows =
+  try
+    let ocamlc_config = if is_windows
+      then "ocamlc.opt.exe -config"
+      else "ocamlc.opt -config"
+    in
+    let config_output = Child_process.execSync ocamlc_config (Child_process.option ~encoding:"utf8" ()) in
+    print_endline ("config_output:\n" ^ config_output);
+
+    let keyvalues = Js.String.split "\n" config_output
+      |> Js.Array.filter (fun x -> Js.String.length x > 0)
+      |> Js.Array.map (fun x ->
+        let index = Js.String.indexOf ":" x in
+        let key = Js.String.substrAtMost ~from:0 ~length:index x in
+        let value = Js.String.substr ~from:(index + 1) x in
+        (Js.String.trim key, Js.String.trim value)
+      )
+    in
+    print_endline "keyvalues";
+    Array.iter (fun (key, value) -> print_endline (key ^ ": " ^ value)) keyvalues;
```

Comment:
can we just use `Js.Array.forEach` here, I would prefer avoid hard dependency on stdlib. Ideally the generated js file should have less deps since once configure failure, everything gets broken